mitiq/factories.py

```diff
@@ -0,0 +1,129 @@
+            outstack = []
+        super(BatchedFactory, self).__init__(instack, outstack)
+        self.scalars = scalars
+
+    def step(self, instack: List[float], outstack: List[float]) -> float:
+        try:
+            next_param = self.scalars[len(outstack)]
+        except IndexError:
+            raise IndexError(f"BatchedFactory cannot take another step. Number of batched scalars ({len(instack)}) "
+                             f"exceeded.")
+        self.instack.append(next_param)
+        return next_param
+
+    def is_converged(self, instack: List[float], outstack: List[float]) -> bool:
+        return len(instack) == len(self.scalars)
+
+# Specific extrapolation algorithms are given below.
+# They are implemented as child classes of Factory or BatchedFactory.
+
+# TODO list:
+# [x] Richardson's extrapolation
+# [x] Linear fit
+# [x] Polynomial fit
+# Log-Richardson's extrapolation
+# Log-Polynomial fit
+# Adaptive Linear fit
+# Adaptive Log-polinomial fit
+# Adaptive max-likelihood
+# Adaptive Bayesian
+# etc... 
+
+
+class RichardsonFactory(BatchedFactory):
+    """Factory object implementing the Richardson's extrapolation algorithm."""
+
+    def reduce(self, x: List[float], y: List[float]) -> float:
+        """ Given two lists of x and y values associated to an unknwn function y=f(x), returns 
+        the extrapolation of the function to the x=0 limit, i.e., an estimate of f(0).
+        The algorithm is based on the Richardson's extrapolation method.
+        """
+
+        def get_gammas(c: float) -> List[float]:
+            """Returns the linear combination coefficients "gammas" for Richardson's extrapolation.
+            The input is a list of the noise stretch factors.
+            """
+            order = len(c) - 1
+            np_c = np.asarray(c)
+            A = np.zeros((order + 1, order + 1))
+            for k in range(order + 1):
+                A[k] = np_c ** k
+            b = np.zeros(order + 1)
+            b[0] = 1
+            return np.linalg.solve(A, b)
+        
+        # check arguments are valid
+        assert len(y) > 0
+        assert len(x) == len(y)
+        # Richardson's extrapolation
+        gammas = get_gammas(x)
+        return np.dot(gammas, y)
+
+
+class LinearFactory(BatchedFactory):
+    """Factory object implementing a zero-noise extrapolation algotrithm based on a linear fit."""
+
+    def reduce(self, x: List[float], y: List[float]) -> float:
+        """ Given two lists of x and y values associated to an unknwn function y=f(x), returns 
+        the extrapolation of the function to the x=0 limit, i.e., an estimate of f(0).
+        The algorithm determines, with a standard least squared method, the parameters 
+        (q, m) such that the line g(x) = q + m*x optimally fits the input points.
+        Returns g(0) = q.
+        """
+        # linear least squared fit
+        _ , q = np.polyfit(x, y, deg=1)
+        return q
+
+
+class PolyFactory(BatchedFactory):
```

Comment:
Are both `LinearFactory` and `PolyFactory` necessary? In principle, `LinearFactory.reduce` is just `PolyFactor.reduce(..., order=1)`. 

Maybe it's good to have them separate conceptually, but implement `LinearFactory` as described above? Ping @willzeng for thoughts.

Reply:
I asked Andrea just the same question! We think alike :) He had a good answer though, which was that it is likely that linear fits have a more optimal algorithm then an order 1 polyfit. The separate class is meant to hold that optimization.

Reply:
Actually also RichardsonFactory is a special case of PolyFactory with `order=length(instack)` !
I would keep different names because I think this is more handy and user friendly. 
However we could change the way in which `LinearFactory` and `RichardsonFactory` are coded.
They could be children of `PolyFactory`, or we could find a way of explicitly calling  `PolyFactory.reduce` in the body of `LinearFactory` and `RichardsonFactory`. I am not sure how to do it this in practice.

Reply:
> Actually also RichardsonFactory is a special case of PolyFactory with `order=length(instack)` !

Ahah!

> I would keep different names because I think this is more handy and user friendly.

Complete agree.

> However we could change the way in which `LinearFactory` and `RichardsonFactory` are coded.

Agree again. Same interface for the user but we can change the implementation.

> They could be children of `PolyFactory`, or we could find a way of explicitly calling `PolyFactory.reduce` in the body of `LinearFactory` and `RichardsonFactory`. I am not sure how to do it this in practice.

I have ideas but maybe not ideal -- open to other suggestions. One idea is to implement a private `_reduce` method in `PolyFactory` which takes in an order (and other arguments). This is then called by the public `reduce` method in `PolyFactory`, and also called by `LinearFactory` with `order=1`.

> I asked Andrea just the same question! We think alike :) He had a good answer though, which was that it is likely that linear fits have a more optimal algorithm then an order 1 polyfit. The separate class is meant to hold that optimization.

For optimization, I think the current approach is good. I'm not sure if the extrapolation part is much of a bottleneck, though.

Reply:
Thanks Ryan, I followed more or less what you suggested. PolyFactory has now a static method `static_reduce` which is called by `PolyFactory`, `LinearFactory`, and `RichardsonFactory`.

Reply:
addressed in #24